File: backend/src/main/java/com/kiwi/features/personality/data/PersonalityDataMapper.java
package com.kiwi.features.personality.data;

import com.kiwi.features.users.data.UsersPersistence;

public class PersonalityDataMapper {

    public static PersonalityDomain toDomain(PersonalityDTO dto) {
        return (dto == null) ? null :
                new PersonalityDomain(
                        dto.getRealName(),
                        dto.getKnightName(),
                        dto.getBuild(),
                        dto.getGoodApps(),
                        dto.getBadApps(),
                        dto.getNeutralApps()
                );
    }

    public static PersonalityDomain toDomain(PersonalityPersistence persistence) {
        return (persistence == null) ? null :
                new PersonalityDomain(
                        persistence.getRealName(),
                        persistence.getKnightName(),
                        persistence.getBuild(),
                        persistence.getGoodApps(),
                        persistence.getBadApps(),
                        persistence.getNeutralApps()
                );
    }

    public static PersonalityPersistence toPersistence(UsersPersistence usersPersistence, PersonalityDomain domain) {
        return (domain == null) ? null :
                PersonalityPersistence.builder()
                        .user(usersPersistence)
                        .realName(domain.getRealName())
                        .knightName(domain.getKnightName())
                        .build(domain.getBuild())
                        .goodApps(domain.getGoodApps())
                        .badApps(domain.getBadApps())
                        .neutralApps(domain.getNeutralApps())
                        .build();
    }

    public static PersonalityPersistence toPersistence(UsersPersistence usersPersistence, PersonalityDTO dto) {
        return toPersistence(usersPersistence, toDomain(dto));
    }

    public static void updatePersistence(PersonalityPersistence persistence, PersonalityDomain domain) {
        persistence.setRealName(domain.getRealName());
        persistence.setKnightName(domain.getKnightName());
        persistence.setBuild(domain.getBuild());
        persistence.setGoodApps(domain.getGoodApps());
        persistence.setBadApps(domain.getBadApps());
        persistence.setNeutralApps(domain.getNeutralApps());
    }

    public static PersonalityDTO toDTO(PersonalityDomain domain) {
        if (domain == null) return null;
        return new PersonalityDTO(
                domain.getRealName(),
                domain.getKnightName(),
                domain.getBuild(),
                domain.getGoodApps(),
                domain.getBadApps(),
                domain.getNeutralApps()
        );
    }

    public static PersonalityDTO toDTO(PersonalityPersistence persistence) {
        return toDTO(toDomain(persistence));
    }
}

File: mobile/app/src/main/java/com/bellako/kiwi/features/personality/data/PersonalityDataMapper.kt
package com.bellako.kiwi.features.personality.data

import android.os.Build
import androidx.annotation.RequiresApi

object PersonalityDataMapper {
    @RequiresApi(Build.VERSION_CODES.O)
    fun toDomain(dto: PersonalityDTO): PersonalityDomain =
        PersonalityDomain(
            realName = UserName.of(dto.realName).getOrNull(),
            knightName = UserName.of(dto.knightName).getOrNull(),
            build = dto.build,
            goodApps = dto.goodApps,
            badApps = dto.badApps,
        )

    @RequiresApi(Build.VERSION_CODES.O)
    fun toDomain(state: PersonalityState): PersonalityDomain =
        PersonalityDomain(
            realName = UserName.of(state.realName).getOrNull(),
            knightName = UserName.of(state.knightName).getOrNull(),
            build = state.build,
            goodApps = state.goodApps,
            badApps = state.badApps,
        )

    fun toState(domain: PersonalityDomain): PersonalityState =
        PersonalityState(
            realName = domain.realName?.value ?: "",
            knightName = domain.knightName?.value ?: "",
            build = domain.build,
            goodApps = domain.goodApps,
            badApps = domain.badApps,
        )

    @RequiresApi(Build.VERSION_CODES.O)
    fun toState(dto: PersonalityDTO): PersonalityState = toState(toDomain(dto))

    fun toDTO(domain: PersonalityDomain): PersonalityDTO =
        PersonalityDTO(
            realName = domain.realName?.value ?: "",
            knightName = domain.knightName?.value ?: "",
            build = domain.build,
            goodApps = domain.goodApps,
            badApps = domain.badApps,
        )

    @RequiresApi(Build.VERSION_CODES.O)
    fun toDTO(state: PersonalityState): PersonalityDTO = toDTO(toDomain(state))
}
